### PR TITLE
Local Code Coverage

### PR DESCRIPTION
**Pull request by freddiejbawden**
## Description
This PR adds the functionality to run a local line coverage analysis using `lcov` and `genhtml`. We currently check the % of statements covered.

 *Click up link*: [https://app.clickup.com/t/2ecn1w](https://app.clickup.com/t/2ecn1w)

## Changes
- split bin folder into `bin/release` and `bin/debug` as coverage requires additional tooling to be added to `.o` files 
- adds `make coverage` option which build and runs the test suite and outputs coverage report

## Additional Info 
We use `genhtml` to produce a human readable version of our coverage data
> Screenshot 2019-11-20 at 13 16 38
> An overall view of the code coverage across the submodules, we have written a small test for the Data class.

> Screenshot 2019-11-20 at 13 16 44
> Each sub directory has a per file break down of % coverage

> Screenshot 2019-11-20 at 13 16 56
> You can see how many times a line was run at great detail!

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -8,7 +8,12 @@ NOLINT=0
 
 SRCS_DIR:=src
 LIBS_DIR:=lib
-OBJS_DIR:=bin
+OBJS_DIR:=bin/
```
Comment (mkristien) [inline Makefile]: This seems like a redundant `/`
```suggestion
OBJS_DIR:=bin
```

File: Makefile [hunk 2]
```diff
@@ -8,7 +8,12 @@ NOLINT=0
...
+OBJS_RELEASE_DIR:=${OBJS_DIR}/release
+OBJS_DEBUG_DIR:=bin/debug
+
+COVERAGE_FLAGS:=--coverage
```
Comment (mkristien) [inline Makefile]: Why not hide this in `utils/config.mk`? Then this setting can be removed from here and from `test/Makefile`.

File: test/utils/get_code_cov.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
+if ![ lcov -v given-command > /dev/null 2>&1 ]; then
```
Comment (mkristien) [inline test/utils/get_code_cov.sh]:
```suggestion
#!/bin/bash
set -e
if ![ lcov -v given-command > /dev/null 2>&1 ]; then
```

Comment (mkristien) [inline test/utils/get_code_cov.sh]: For better bash resolution of "does command exist?"
```suggestion
if [ -z `which lcov` ]; then
```

**Review by mkristien — CHANGES_REQUESTED** [review]
Good work :)

I struggle to rerun this on my machine. Which version of g++, lcov, gcov, did you use?

File: test/utils/get_code_cov.sh (hunk 3 above)
Comment (freddiejbawden) [reply test/utils/get_code_cov.sh]: What does `set -e` do?

File: Makefile (hunk 2 above)
Comment (freddiejbawden) [reply Makefile]: Have moved `OBJS_DEBUG_DIR` and `COVERAGE_FLAGS` to Config, I have kept `OBJS_RELEASE_DIR` in ./Makefile as it is only used there

File: test/utils/get_code_cov.sh (hunk 3 above)
Comment (freddiejbawden) [reply test/utils/get_code_cov.sh]: Resolved through some googling, updated `get_code_cov.sh`

**Comment by freddiejbawden** [comment]
> Good work :)
> 
> I struggle to rerun this on my machine. Which version of g++, lcov, gcov, did you use?

```
g++ -> 4.2.1
lcov: LCOV version 1.14
```
Installed through `brew`

Admittly haven't tested on anything but OSX yet, so might need some X-platform ironing out

File: Makefile (hunk 2 above)
Comment (freddiejbawden) [reply Makefile]: Have moved `OBJS_DEBUG_DIR` and `COVERAGE_FLAGS` to `config.mk` 👍

**Review by freddiejbawden — COMMENTED** [review]
Not sure how i opened this review... please ignore

**Comment by mkristien** [comment]
I am happy with this. This is just an experimental feature for now. It does not need to be perfect as it does not effect other people's development. We want to get it out there so that people can test it.

@freddiejbawden , you could have changed the other if condition in `test/utils/get_code_cov.sh` as well. Merge when you are ready.

**Review by mkristien — APPROVED** [review]

**Comment by neilweidinger** [comment]
Hang on just pushed a commit, @mkristien see if it's still ok pls

**Review by mkristien — APPROVED** [review]

**Pull request merged**